File: src/main/java/innexgo/hours/SchoolDataService.java
/*
 * Innexgo Website
 * Copyright (C) 2020 Innexgo LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package innexgo.hours;

import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class SchoolDataService {

  @Autowired
  private JdbcTemplate jdbcTemplate;

  public SchoolData getBySchoolDataId(long schoolDataId) {
    String sql = "SELECT * FROM school_data WHERE school_data_id=?";
    RowMapper<SchoolData> rowMapper = new SchoolDataRowMapper();
    SchoolData school = jdbcTemplate.queryForObject(sql, rowMapper, schoolDataId);
    return school;
  }

  public long nextId() {
    String sql = "SELECT max(school_data_id) FROM school_data";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if (maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }

  public void add(SchoolData schoolData) {
    schoolData.schoolDataId = nextId();
    schoolData.creationTime = System.currentTimeMillis();
    // Add schoolData
    String sql = "INSERT INTO school_data values (?, ?, ?, ?, ?, ?, ?)";
    jdbcTemplate.update(sql, schoolData.schoolDataId, schoolData.creationTime, schoolData.creatorUserId,
        schoolData.schoolId, schoolData.name, schoolData.description, schoolData.active);
  }

  public boolean existsBySchoolDataId(long schoolId) {
    String sql = "SELECT count(*) FROM school WHERE school_id=?";
    long count = jdbcTemplate.queryForObject(sql, Long.class, schoolId);
    return count != 0;
  }

  public SchoolData getBySchoolId(long schoolId) {
    return query( //
        null, //
        null, //
        null, //
        null, //
        null, //
        schoolId, //
        null, //
        null, //
        null, //
        null, //
        null, //
        true, //
        null, //
        0, //
        1).findFirst().orElse(null);
  }

  public Stream<SchoolData> query( //
      Long schoolDataId, //
      Long creationTime, //
      Long minCreationTime, //
      Long maxCreationTime, //
      Long creatorUserId, //
      Long schoolId, //
      String name, //
      String partialName, //
      String description, //
      String partialDescription, //
      Boolean active, //
      boolean onlyRecent, //
      Long recentAdminUserId, //
      long offset, //
      long count) //
  {

    boolean nojoinrecentadminship = recentAdminUserId == null;

    String sql = "SELECT sd.* FROM school_data sd" + (!onlyRecent ? ""
        : " INNER JOIN (SELECT max(school_data_id) id FROM school_data GROUP BY school_id) maxids ON maxids.id = sd.school_data_id")
    // first join all adminships
        + (nojoinrecentadminship ? "" : " INNER JOIN adminship a ON a.school_id = sd.school_id")
        // then filter adminships based on recent
        + (nojoinrecentadminship ? ""
            : " INNER JOIN (SELECT max(adminship_id) id FROM adminship GROUP BY user_id, school_id) maxadminids ON maxadminids.id = a.adminship_id")
        + " WHERE 1=1 " + (schoolDataId == null ? "" : " AND sd.school_data_id = " + schoolDataId)
        + (creationTime == null ? "" : " AND sd.creation_time = " + creationTime)
        + (minCreationTime == null ? "" : " AND sd.creation_time > " + minCreationTime)
        + (maxCreationTime == null ? "" : " AND sd.creation_time < " + maxCreationTime)
        + (creatorUserId == null ? "" : " AND sd.creator_user_id = " + creatorUserId)
        + (schoolId == null ? "" : " AND sd.school_id = " + schoolId)
        + (name == null ? "" : " AND sd.name = " + Utils.escape(name))
        + (partialName == null ? "" : " AND sd.name LIKE " + Utils.escape("%" + partialName + "%"))
        + (description == null ? "" : " AND sd.description = " + Utils.escape(description))
        + (partialDescription == null ? "" : " AND sd.description LIKE " + Utils.escape("%" + partialDescription + "%"))
        + (active == null ? "" : " AND sd.active = " + active)
        + (recentAdminUserId == null ? "" : " AND a.user_id = " + recentAdminUserId)
        + (recentAdminUserId == null ? "" : " AND a.adminship_kind = " + AdminshipKind.ADMIN.value)
        + (" ORDER BY sd.school_data_id") + (" LIMIT " + offset + ", " + count) + ";";

    RowMapper<SchoolData> rowMapper = new SchoolDataRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper).stream();
  }

}

File: rust/src/school_data_service.rs
use super::db_types::*;
use super::utils::current_time_millis;
use std::convert::From;
use tokio_postgres::GenericClient;

impl From<tokio_postgres::row::Row> for SchoolData {
  // select * from school_data order only, otherwise it will fail
  fn from(row: tokio_postgres::Row) -> SchoolData {
    SchoolData {
      school_data_id: row.get("school_data_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      school_id: row.get("school_id"),
      name: row.get("name"),
      description: row.get("description"),
      active: row.get("active"),
    }
  }
}

pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  school_id: i64,
  name: String,
  description: String,
  active: bool,
) -> Result<SchoolData, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let school_data_id = con
    .query_one(
      "INSERT INTO
       school_data(
           creation_time,
           creator_user_id,
           school_id,
           name,
           description,
           active
       )
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING school_data_id
      ",
      &[
        &creation_time,
        &creator_user_id,
        &school_id,
        &name,
        &description,
        &active,
      ],
    )
    .await?
    .get(0);

  Ok(SchoolData {
    school_data_id,
    creation_time,
    creator_user_id,
    school_id,
    name,
    description,
    active,
  })
}

pub async fn get_by_school_data_id(
  con: &mut impl GenericClient,
  school_data_id: &i64,
) -> Result<Option<SchoolData>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM school_data WHERE school_data_id=$1",
      &[&school_data_id],
    )
    .await?
    .map(|x| x.into());

  Ok(result)
}

pub async fn get_by_school_id(
  con: &mut impl GenericClient,
  school_id: i64,
) -> Result<Option<SchoolData>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "
      SELECT sd.* FROM school_data sd
      INNER JOIN (
          SELECT max(school_data_id) id
          FROM school_data
          GROUP BY school_id
      ) maxids
      ON maxids.id = sd.school_data_id
      WHERE sd.school_id = $1
      ",
      &[&school_id],
    )
    .await?
    .map(|x| x.into());
  Ok(result)
}

pub async fn is_active_by_school_id(
  con: &mut impl GenericClient,
  school_id: i64,
) -> Result<bool, tokio_postgres::Error> {
  let result = match get_by_school_id(con, school_id).await? {
    Some(SchoolData { active: true, .. }) => true,
    _ => false,
  };

  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: innexgo_hours_api::request::SchoolDataViewProps,
) -> Result<Vec<SchoolData>, tokio_postgres::Error> {

  let sql = [
    "SELECT sd.* FROM school_data sd",
    if props.only_recent {
      " INNER JOIN
          (SELECT max(school_data_id) id FROM school_data GROUP BY school_id) maxids
          ON maxids.id = sd.school_data_id"
    } else {
      ""
    },
    " WHERE 1 = 1",
    " AND ($1::bigint[] IS NULL OR sd.school_data_id IN $1)",
    " AND ($2::bigint   IS NULL OR sd.creation_time >= $2)",
    " AND ($3::bigint   IS NULL OR sd.creation_time <= $3)",
    " AND ($4::bigint   IS NULL OR sd.creator_user_id = $4)",
    " AND ($5::bigint   IS NULL OR sd.school_id = $5)",
    " AND ($6::text     IS NULL OR sd.name = $6)",
    " AND ($7::text     IS NULL OR sd.name LIKE CONCAT('%',$7,'%'))",
    " AND ($8::text     IS NULL OR sd.description = $8)",
    " AND ($9::text     IS NULL OR sd.description LIKE CONCAT('%',$9,'%'))",
    " AND ($10::bool    IS NULL OR sd.active = $10)",
    " ORDER BY sd.school_data_id",
    " LIMIT $11",
    " OFFSET $12",
  ]
  .join("");

  let stmnt = con.prepare(&sql).await?;

  let results = con
    .query(
      &stmnt,
      &[
        &props.school_data_id,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.school_id,
        &props.name,
        &props.partial_name,
        &props.description,
        &props.partial_description,
        &props.active,
        &props.count.unwrap_or(100),
        &props.offset.unwrap_or(0),
      ],
    )
    .await?
    .into_iter()
    .map(|row| row.into())
    .collect();

  Ok(results)
}
